File: cpp/1690_stone_game_vii/stone_game_vii.cpp
/*
 * @Date: 2024-02-03
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-02-03
 * @FilePath: /algorithm/cpp/1690_stone_game_vii/stone_game_vii.cpp
 */

#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int stoneGameVII(vector<int>& stones) {
    int n = stones.size();
    vector<int> sum(n + 1);
    vector<vector<int>> dp(n, vector<int>(n, 0));

    for (int i = 0; i < n; i++) {
      sum[i + 1] = sum[i] + stones[i];
    }
    for (int i = n - 2; i >= 0; i--) {
      for (int j = i + 1; j < n; j++) {
        dp[i][j] = max(sum[j + 1] - sum[i + 1] - dp[i + 1][j], sum[j] - sum[i] - dp[i][j - 1]);
      }
    }

    return dp[0][n - 1];
  }
};

int main() {
  vector<tuple<vector<int>, int>> tests{
      {{5, 3, 1, 4, 2}, 6},
      {{7, 90, 5, 1, 100, 10, 10, 2}, 122},
  };

  for (auto& [stones, ans] : tests) {
    assert(Solution().stoneGameVII(stones) == ans);
  }
}

File: rust/1690_window.native/stone_game_vii.rs
/*
 * @Date: 2024-02-03
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-02-04
 * @FilePath: /algorithm/rust/1690_window.native/stone_game_vii.rs
 */

struct Solution;

impl Solution {
    pub fn stone_game_vii(stones: Vec<i32>) -> i32 {
        let n = stones.len();
        let mut s = vec![0; n + 1];
        for (i, &x) in stones.iter().enumerate() {
            s[i + 1] = s[i] + x;
        }
        let mut f = vec![0; n];
        for i in (0..n - 1).rev() {
            for j in i + 1..n {
                f[j] = (s[j + 1] - s[i + 1] - f[j]).max(s[j] - s[i] - f[j - 1]);
            }
        }
        f[n - 1]
    }
}

fn main() {
    let tests = vec![
        (vec![5, 3, 1, 4, 2], 6),
        (vec![7, 90, 5, 1, 100, 10, 10, 2], 122),
    ];

    for (stones, ans) in tests {
        assert_eq!(Solution::stone_game_vii(stones), ans);
    }
}
